### ISSUE TITLE
Bug: Updates to keyed lists break FLIP animations when they occur mid-animation

### ISSUE THREAD

**Issue by dead-claudia** [facebook/react#19406]
React version: 16.13.1

## Steps To Reproduce

1. Open code example below
2. Observe animation

Link to code example:

[Flems](https://flems.io/#0=N4Igxg9gdgzhA2BTEAucD4EMAONEBMQAaEAMwEskZUBtUKTAW2TQDoALAF0fmPSk6IBqEAB4ARgFdOnaAAJy+ALwAdELIDmGpGoB8AFQhakogPRSZ0XSqij85AG5yYnAJ5JVIezGxZXKUiQADz0bOTlRSXgFZTVsACdEBz1ReHJdAAYzNN1U9IBGbPS83QAmItycgGYKkoAWWpyAVkb0gDZW3QB2WtMo6yhwyOjFT0wockZMQRS++AGhqJjPKEQgzhScrNMcksKd4pzyg8r0mpP6zpaLnI6b9J6LuYGzewddPjwkME5yaGo0BkUBkALRNFBtEAAXyI9CYLBArAAVtQSJABEJOCJ0S45ONJtNEAAlCAQThyJRyfAQMCSZgCVgaRCcACiSHpnAAQq4AJL4AAUanxUxmIAAlDYkOSHJh4JJEDAKXIAILxeKYVysUjxCCMfnAJBQDScdgoOQADhhcn5AH0iAoxRTdA6bDZSJIoD8-oNElB8Ih4vzHcAwnIiYhMD8ACIAeQAsqxff7A6HwjK5QrWFNsPzyE6wxGfqwwIlCWzEBzBSA0mp7cAANaIfwKK3kMVioipvETEXE0kbQZyCVQKGuqBShQY+LppV4Tj6SaICDSfmSbD4QmciD4Vz2poZDLDt0er3yNcbwRbndBuQhwfkKczylzhfMZecVfrzfb3dyfeH0McXJBIkiVdN5RgUNTFMOQuE4XAUGgoRWAAd3IetyGwAhyEwVgIHiDRTDQjDTAAMXIGB2ADABSFlSmo80MmogBOKoAE1CRgG1KMkUhAkQQD-nJVZ1jA2UIKVECHFYGA0jARAg1DUh8OtCc80pETOFYQ1jXYOQQTkfIAG4FDkZ0MhM8gQRBYMuyAuQkSVONpnYLV4FJQNnJNRNxmpPVHQAKmtPMAGpDPbOyhLkQRGGwJVNJoJEAF0uwS5L4rWTgaHIFLB3CBKcqVGLsFDUcoFDRIAEcIM4ZUe2mb1SPVZh+RvJRnTvcJwnsqSABkKPJSlqVpDlGWZcsOW5PkqyktQxWLdhKHwAA5bcFUi2BhMy-rcSGmk6UxMbWXZTEpoFNRNLmhaltW-1IPKvK5Gggt8OTOREBOgRFQfORYzjLtlPiVTmVMykLNM0Q5Ck7ShF0yyQpC2zHvCPqBuypLWEEdYAGFoEEARJMSBx0a7MqurkQHgfJdS5HBvNIc0mGjRNeHEdvLt8u2tGcsxzLcYxAmNMyknHrKjbcSkibFUpVV1U1bVdX5YVCRJMlrvgFa1pgYdya7Z7yPiFxxfJChDc4AAFTBfUG7sCUEVWtKZLllz9B8NGxtJMXDH5FMe+zTd2qGialrMcH5RB8068m5HssBFo1pVECOrcPXsI0PfIL3EB9nXo8SThJHiQYo+j6KIGwM046WzHy-0inyDNy3rZr7BO2RrqkFIThK-j-AYa7uuA4tq3Ds7zg29LsmuqhI9HqTANfb1mDeswXEwp5KAHADAdyfsrAXCbzElWV+3+2Tl20-dz2BG9j9c+6qLGAgLf8ClwmkhDih4EEQNw-tNtI4cxjlFfeNsk5OxTq7dO19OC319qXey+AghKiHujfu5IDL8lAeguuoDD4MjHvfLqiDXAoIbi4NBsg4qYOwVQ3Bq9h7NyoUQ8I5BSDWiQRSJQYM5AADJeFUlIdwsGjp86F0GKQWUeAgFJxcO4JOnB1SwEBowJUAADRR4xZKEn5AAEmAEgqE2AghEH0TuIxQQxRqJkTJNwSBMZKJgOQX40AoyFwavISkagMiQRAEAsRRdorxHlKTWe5N9b4XknIRIII-Dvi7MNA6DJxA-jwnxOcAAJRA5ANBcDHOEmC5s-Bdifi-T++EWSRnYK1RA-9HTtXZu3WOsp4DiEjPWJUbUOpAPCEnRIpTEAsi3gIHa+MF5qE0bAZx3ohD4FrDHFpbSwD1ntJI+AeAWFdSTmAfeMBRmJgrM-BSEzHHTOgHNIBU9ybbN2fszA+BzrqFOS48q4obH3NfsMzgoyhDjKeVos5qw-TzLAIs9pqypGIE2bIuxCjHEqMTrY+RDiAUvLceqF5SoADkWLQmlWHGLP0+1RpOwmqdXkAosWaG0IgLF81oA7PIMszp9Tun3nYbmR8soka7yQFbV8S4VwPh-umIhwqAxPjkFAKI8BSrvXWRHEuchzzfmvLnKeOYqVGBpVi-+XLogAEINIyoJWOeegYxR8EgLFSgAZUDSvgPAGEcJmAiGLDAVE-B8ZYjYJMpxmKo5+sBWaUoio-UqKMjYKEnwPrZxeQCEA+QUD5CaNCJKJA0hQHrACOgIAGCurQKWIs-ot7uWwKNFEfBC68DQHBBC0EPTYHrBoYsupTBFs4AAAXyG0Vg+Qqh9r6IwfA7bCxaRLR9cuFbPVuCwiIGAJZMJYmdXm+EIgO0gj8qwCdZbp1VviDWkAdaYCIT6FAJtLbrWjsjJwTdupu29v7YOukI6N1bp3VOw6laSCzoRAu+IS601QiAA)

## The current behavior
Animations are choppy

## The expected behavior
Animations are smooth like in [this example](https://flems.io/#0=N4Igxg9gdgzhA2BTEAucD4EMAONEBMQAaEAMwEskZUBtUKTAW2TQDoALAF0fmPSk6IBqEAB4ARgFdOnaAAJy+ALwAdELIDmGpGoB8AFQhakogPRSZ0XSqij85AG5yYnAJ5JVIezGxZXKUiQADz0bOTlRSXgFZTVsACdEBz1ReHJdAAYzNN1U9IBGbPS83QAmItycgGYKkoAWWpyAVkb0gDZW3QB2WtMo6yhwyOjFT0wockZMQRScrNMcksKF4pzylcr0mo36zpadnI6D9J6dvvgBoaiYzyhEIM5Z9PnFnOXX9PWP3W3vhuPdPtvkdvqdFucBmZ7A5dHw8EgwJxyNBqGgqih8hk6iAAL5EehMFggVgAK2oJEgAiEnBElJccnGk2miAAShAIJw5Eo5PgIGBJMwBKwNIhOABRJCCzgAIVcAEl8AAKNSMqYzEAAShsSE5YEwYHYBAliEYMC5cgAgvF4phXKxSPEIIxFarmWyOawDZR8AA5CD4RAwLVQHUKKnxByYaLcvCcfSTRAQaSKyTYfDM6X+1xEORNDIZYM2UiSKCI5GDRJQAPxRUauTAMJyUgQeJyRV0zmISVyCCkOR6g1GyVBhkTNWs9mcVg4bBCfAAYXY3sVXZNwZxNiLJbL8lT6cEmfwrlr9cb5HDkejzlF8eYSc4KbTGazObzBc3g37KM5CSSxtN5oDoa+D-jAjYdnIdwPKBgH6sBMHcr+DigawMBpGAiC1o2pimHIXCcLgKA4UIrAAO7kAA1uQs72JgrAthopjkVRpgAGLkDAhrxAApGKpTcQAHBk3EAJxVAAmsyMAAPqcZIpCBIgjbNq2iqhuQ5pQeKw6sEgUAaJw7ByAAtHI+QANwKHIuhyBklnkMZxl1g2n7hBBJLmgAstM7D2vA7I1t5hmsDaVZOieABUbYaQA1GZGrBuEbnfnIgiMNgmn3NpJowDQJIALqNuEWmgXl+WZdBw40OQhWuZBWWlTV5ppdgjYblAjaJAAjpIgacBaY7TOWrE2swionkoNkuUlX6wD+iQOAAMhxnLcry-JSsKor-tSsoKsqIBIWoGqeku8C+v6gZFbN9Jact9JrXyArUlt2VSntSpqFpx2nd6foBmBHV1Thcgsom8TVnIq5Sma55yAAIgA8p510qW26nmnZVmiHISEoXpBnsPZMUxc513hEh92cNV+WsIIDzztAggCOaeNVTVdNZYzVICNd7Woy26OilZ3JYxpOMlTpBOGcTpOnnVxVZVTNOcwzTPUhVb25Rz9OcNzzOPHV-PA7h7HxC410QRQ5ucAACpglaraOTKCO6U4ijKSZVueGjzmk1Jg4iWF1Vb5A26zC0oVM2ArlyU3kzdupnfg5qIK9mYlvY+l++QAeIEHiUzXIiScJI8SDNNRfhLI2AoP2yd0xAGWmdbLj247jfYEQCfhEgpCcHXXrnbpiD9yZTZh23DsvX3nDdwrcjtTNOKFnVlbVsH10g4tmD0nFcpQA4iDxIbM0QVgU+O+arqu5O6de1nvv+wIgcPoXyVzXIjAQEfIHDhHf4dIUHgIIGsK4czkDrJNeWVdz67ydmnD2GdvbZ2fpwV+wdYEpXwEEc0rdqYc1nuPNS8CR5j1MhfO208hSz3fklCCR48GTwIbTGuxDKGd3HpQ9uL0a50PCOQPsiocFciUKLOQAAyCRPJXCiNFnWEuZdBikCjHgBOacXDuDTpwUKMAVKMHNAAAx0eMNCzJFQABJgA4JxNgIIRArFHlsUEDUhj1GoTcEgOmujyBImgPDMuQ15DcjUBkMCIAE6KPLqleIvU+arxmiDViLYMLF0QMZPw95rrrWekKcQWZ6IKVjAACUQOQDQXAPyJNwrbPw11v6-xQipMUcFxqIAgVA+OC8IJ6ngPAcQ+oKLmgml0qu4Q06JAaYgMUR8BBUyEMfA6JjYC+PLHONQOZen9MGTmFR8A8D8KSmnMAF8YBUxCiaH+mE1DLJgKs6Ax0E5LyLsc055zMD4E+uoHxfiOqancR8kCszODzLuDWG5Py1lVg2f2KM2ywAUV2aoxAhyNGeO0bo-RqcPFaO8aY+5UAAk2l+eaAA5KS+JbV1wfhBgAVSRGkNwchxC+MBjkzaHsdoCA+oqUlmhtCIFJSdaAJzyAIuGZ0mBCghHnlAZeMmIckAO1vImZMsrj7yuuuqiMUZNJRHgG1KG+zEBSvCHuZ8R5MGL0bDHPlRgBWkogReXVABCbkUB9XUqBuvRZwY+CQHSpQY+IgBniC7LifEIAGDMBEJ6GA5J+AG1jbcglUqU2-LrqUM0tz9HmRsDiOEq4dywBEPkDETRcT5RxEAA)

-----

Context: this bug is far from unique to React: https://twitter.com/isiahmeadows1/status/1284726730574315522

**Edit:** Also relevant: https://github.com/whatwg/html/issues/5742

**Comment by rickhanlonii** [comment]
Hey @isiahmeadows, thanks for filing this issue!

Can you reduce this to a specific React bug? As is, the complaint seems to be that the code just doesn't work as expected. That can happen for a number of reasons that are not bugs in React, so what's the specific React bug we should investigate?

**Comment by isiahmeadows** [comment]
@rickhanlonii The specific issue is that React applies a keyed diff in such a way that it results in the choppy animations. The only functional difference between the React snippet and the vanilla snippet linked in the "expected behavior" section is the render function. They're otherwise doing the same exact thing to the DOM, and I've verified this through DOM inspection. I've also narrowed it down to `appendChild` vs `insertBefore`, as referenced in the linked spec bug.

It's not a bug in React *itself*, but a browser bug that React needs to work around.
